iis/mymodule.cpp

```diff
@@ -467,6 +407,204 @@ HRESULT CMyHttpModule::ReadFileChunk(HTTP_DATA_CHUNK *chunk, char *buf)
+
+    // Remove/Modify Transfer-Encoding header if "chunked" Encoding is set in the request. 
+    // This is to avoid sending both Content-Length and Chunked Transfer-Encoding in the request header.
+    static const std::string CHUNKED = "chunked";
+    USHORT encodingLength = 0;
+    const char* transferEncoding = iisRequest->GetHeader(HttpHeaderTransferEncoding, &encodingLength);
+    if (transferEncoding)
+    {
+        if (CHUNKED.size() == encodingLength &&
+            std::equal(CHUNKED.cbegin(), CHUNKED.cend(), transferEncoding,
+                [](char lhs, char rhs) { return std::tolower(lhs) == std::tolower(rhs); }))
+        {
+            iisRequest->DeleteHeader(HttpHeaderTransferEncoding);
+        }
+    }
+
+    auto contentLengthStr = std::to_string(contentLength);
+    HRESULT hr = iisRequest->SetHeader(
+        HttpHeaderContentLength,
+        contentLengthStr.c_str(),
+        contentLengthStr.size(),
+        TRUE);
+
+    if (FAILED(hr))
+    {
+        return hr;
+    }
+
+    // since we clean the APR pool at the end of OnSendRequest, we must get IIS-managed memory chunk
+    //
+    char* requestBuffer = static_cast<char*>(rsc->httpContext->AllocateRequestMemory(contentLength));
+    status = apr_brigade_flatten(brigade, requestBuffer, reinterpret_cast<apr_size_t*>(&contentLength));
+    if (status != APR_SUCCESS)
+    {
+        return E_FAIL;
+    }
+    return iisRequest->InsertEntityBody(requestBuffer, contentLength);
+}
+
+
+static apr_status_t ReadBodyCallback(request_rec *r, char *buf, unsigned int length, unsigned int *readcnt, int *is_eos)
+{
+    RequestStoredContext *rsc = RetrieveIISContext(r);
+
+    *readcnt = 0;
+
+    if (rsc == NULL)
+    {
+        *is_eos = 1;
+        return APR_SUCCESS;
+    }
+
+    IHttpContext *pHttpContext = rsc->httpContext;
+    IHttpRequest *pRequest = pHttpContext->GetRequest();
+
+    if (pRequest->GetRemainingEntityBytes() == 0)
+    {
+        *is_eos = 1;
+        return APR_SUCCESS;
+    }
+
+    HRESULT hr = pRequest->ReadEntityBody(buf, length, false, (DWORD *)readcnt, NULL);
+
+    if (FAILED(hr))
+    {
+        // End of data is okay.
+        if (ERROR_HANDLE_EOF != (hr & 0x0000FFFF))
+        {
+            // Set the error status.
+            rsc->provider->SetErrorStatus(hr);
+        }
+
+        *is_eos = 1;
+    }
+
+    return APR_SUCCESS;
+}
+
+static apr_status_t WriteBodyCallback(request_rec *r, char *buf, unsigned int length)
+{
+    RequestStoredContext *rsc = RetrieveIISContext(r);
+
+    if (!rsc || !rsc->requestRec)
+        return APR_SUCCESS;
+
+    IHttpContext *pHttpContext = rsc->httpContext;
+    IHttpRequest *pHttpRequest = pHttpContext->GetRequest();
+
+    CHAR szLength[21]; //Max length for a 64 bit int is 20
+
+    ZeroMemory(szLength, sizeof(szLength));
+
+    HRESULT hr = StringCchPrintfA(
+        szLength,
+        sizeof(szLength) / sizeof(CHAR)-1, "%d",
+        length);
+
+    if (FAILED(hr))
+    {
+        // not possible
```

Comment:
Yes, makes sense. This code has been moved around but otherwise is not altered compared to the initial version, but definitely no harm improving it. Will do in the next revision.

Reply:
I preferred to remove this chunk of code altogether and use `std::to_string()` instead.
This is simpler, safer and cleaner.